schema-registry/core/src/main/java/com/hortonworks/registries/schemaregistry/DefaultSchemaRegistry.java

```diff
@@ -494,16 +498,63 @@ public Integer getSchemaVersion(String schemaName, String schemaText) throws Sch
         }
 
         Long schemaMetadataId = schemaMetadataInfo.getId();
-        Integer result = findSchemaVersion(schemaMetadataInfo.getSchemaMetadata().getType(), schemaText, schemaMetadataId);
+        SchemaVersionStorable result = findSchemaVersion(schemaMetadataInfo.getSchemaMetadata().getType(), schemaText, schemaMetadataId);
 
         if (result == null) {
             throw new SchemaNotFoundException("No schema found for schema metadata key: " + schemaName);
         }
 
-        return result;
+        return result.getVersion();
     }
 
-    private Integer findSchemaVersion(String type,
+    @Override
+    public SchemaNameVersionId getSchemaNameVersionId(String schemaName, String schemaText) throws SchemaNotFoundException, InvalidSchemaException {
+        SchemaMetadataInfo schemaMetadataInfo = getSchemaMetadata(schemaName);
+        if (schemaMetadataInfo == null) {
+            throw new SchemaNotFoundException("No schema found for schema metadata key: " + schemaName);
+        }
+
+        Long schemaMetadataId = schemaMetadataInfo.getId();
+        SchemaVersionStorable result = findSchemaVersion(schemaMetadataInfo.getSchemaMetadata().getType(), schemaText, schemaMetadataId);
+
+        if (result == null) {
+            throw new SchemaNotFoundException("No schema found for schema metadata key: " + schemaName);
+        }
+
+        return new SchemaNameVersionId(result.getName(), result.getVersion(), result.getId(), result.getSchemaText());
+    }
+
+    public Long registerSchema(String schemaName, String schemaText, String description)
```

Comment:
addSchemaVersion semantics are quite a bit different (e.g. it throws exception if schema name is not found, where as register we have to auto create), already adjusted private dependent methods like findSchemaVersion to return more shareable response objects, to avoid duplication there.